stdlib/public/runtime/BuiltinProtocolConformances.cpp

```diff
@@ -0,0 +1,810 @@
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Definitions of some builtin protocol witnesses.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/BuiltinProtocolConformances.h"
+#include "swift/Runtime/Casting.h"
+#include "swift/Runtime/Debug.h"
+#include "swift/Runtime/Metadata.h"
+
+#include <vector>
+
+using namespace swift;
+
+using StaticInfixWitness = SWIFT_CC(swift) bool(OpaqueValue *, OpaqueValue *,
+                                                SWIFT_CONTEXT const Metadata *,
+                                                const Metadata *,
+                                                const WitnessTable *);
+
+//===----------------------------------------------------------------------===//
+// Tuple Equatable Conformance
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__)
+// Create a GOT equivalent for the Equatable reference.
+__asm(
+  "  .type got." EQUATABLE_DESCRIPTOR_SYMBOL ", @object\n"
+  "  .section .data.rel.ro\n"
+  "  .p2align 3\n"
+  "got." EQUATABLE_DESCRIPTOR_SYMBOL ":\n"
+  "  .quad (" EQUATABLE_DESCRIPTOR_SYMBOL ")\n"
+  "  .size got." EQUATABLE_DESCRIPTOR_SYMBOL ", 8\n"
+);
+
+// Create a GOT equivalent for the Equatable.== method descriptor.
+__asm(
+  "  .type got." EQUATABLE_EE_METHOD_DESCRIPTOR ", @object\n"
+  "  .p2align 3\n"
+  "got." EQUATABLE_EE_METHOD_DESCRIPTOR ":\n"
+  "  .quad (" EQUATABLE_EE_METHOD_DESCRIPTOR ")\n"
+  "  .size got." EQUATABLE_EE_METHOD_DESCRIPTOR ", 8\n"
+);
+#endif
+
+// Define the conformance descriptor for tuple Equatable. We do this in
+// assembly to work around relative reference issues.
+__asm(
+  #if defined(__ELF__)
+  "  .type __swift_tupleEquatable_private, @object\n"
+  "  .local __swift_tupleEquatable_private\n"
+  "  .comm __swift_tupleEquatable_private, 128, 16\n"
+  "  .protected " TUPLE_EQUATABLE_CONF "\n"
+  "  .type " TUPLE_EQUATABLE_CONF ", @object\n"
+  "  .section .rodata\n"
+  #elif defined(__MACH__)
+  "  .zerofill __DATA, __bss, __swift_tupleEquatable_private, 128, 4\n"
+  "  .section __TEXT, __const\n"
+  #endif
+  "  .globl " TUPLE_EQUATABLE_CONF "\n"
+  "  .p2align 2\n"
+  TUPLE_EQUATABLE_CONF ":\n"
+  #if defined(__ELF__)
+  // This is an indirectable relative reference to the GOT equivalent for the
+  // Equatable protocol descriptor, hence why we add 1 to indicate indirect.
+  "  .long (got." EQUATABLE_DESCRIPTOR_SYMBOL " - \
+              (" TUPLE_EQUATABLE_CONF ")) + 1\n"
+  #elif defined(__MACH__)
+  "  .long " EQUATABLE_DESCRIPTOR_SYMBOL "@GOTPCREL + 5\n"
```

Comment:
We should be able to use a direct reference to EQUATABLE_DESCRIPTOR_SYMBOL without going through the GOT, since it is in the standard library and the runtime and standard library are always linked together. Same with the other symbols you generate GOT equivalents for here. You can also use `.` to refer to the current address in assembly language instead of doing the offset math yourself. So this could be:

```
TUPLE_EQUATABLE_CONF ":\n"
"  .long " EQUATABLE_DESCRIPTOR_SYMBOL " - .\n"
```

And that should also be portable between ELF and Mach-O targets as well.

If you do end up needing GOT entries, though, IIRC the syntax in ELF for a GOTPCREL reference is `symbolname@GOT`; you shouldn't need to make your own GOT equivalent variable.

Reply:
That's what I thought too when I initially wrote this, but stumbled upon `'symbol' can not be undefined in a subtraction expression` for the stdlib symbols.

Regarding ELF `@GOT`: Interesting! Any reason LLVM doesn't emit this for linux binaries? I know LLVM is smart enough to use the Darwin syntax when it encounters those GOT entries.

As I'm compiling this after rebasing from upstream, it seems the build-script also wants to build the arm64 stdlib on my x86_64 mac, which leads to a linker error (of course!) `illegal text-relocation to 'symbol'` for the `==` method descriptor reference (I'm curious why it complains for that symbol and not the protocol descriptor?) Any idea how I can fix this, or debug on my machine? After some very basic research, maybe it's referring to absolute addressing on arm64 not being allowed? Although if that were the case then the protocol descriptor would error out too... I can only assume it's also going to error out with the other method descriptors as well.

Reply:
Looking at llvm source, it looks like aarch64 (aka arm64, right?) doesn’t support GOT pc rel annotation, but it does support `foo@GOT - .`... I’ll do some more investigating once I’m off work.

Reply:
I did some reading, and it turns out ELF has `@GOTPCREL` as well! It differs from macho in that the required `+ 4` is not required, so I can do `symbol@GOTPCREL + 1` for an indirect indirectable pointer. On ARM, it looks like `@GOT - .` is the equivalent from what I've read (which also compiles), but I haven't been able to test it yet. Even though these symbols require going through GOT, I can still take advantage of cleaning up the math by using `- .` in places where GOT is not needed. Please take a look at the latest commit and let me know what you think!

Reply:
Ah, right, LLVM needs an upstream fix for x86-64 to allow for PC-relative references to undefined symbols. This should be OK then.